squadutilslib.py

```diff
@@ -0,0 +1,263 @@
+        output_filename,
+    )
+    return testrun
+
+
+def find_build(build_names, build_options, suite_names, envs, project):
+    """
+    Given a list of builds IDs to choose from in a project, find the first one
+    that has a match for the build name, suite names and environments
+    """
+    build = None
+    build_name = None
+    testrun = None
+    suites = None
+
+    # Given a list of builds IDs, find one that contains the suites and that
+    # has a matching build name
+    for build_option in build_options.values():
+        # only pick builds that are finished
+        if not build_option.finished:
+            logger.info(f"Skipping {build_option.id} as build is not marked finished")
+            continue
+        logger.info(f"Checking build {build_option.id}")
+        # Create the list of suite IDs from the suite names
+        if suite_names:
+            suites = []
+            for s in suite_names:
+                suites += project.suites(slug=s).values()
+
+        # Download the tests for the current build option
+        testrun = download_tests(
+            project=project,
+            build=build_option,
+            build_names=build_names,
+            suites=suites,
+            envs=envs,
+        )
+        if not testrun:
+            logger.info(f"No tests found that matches the criteria in build ID {build_option.id}")
+            continue
+        else:
+            break
+
+    if not testrun:
+        logger.debug(f"No test for {project.id}-{build_option.id}-{suites}-{envs}-{build_option.version}")
+    else:
+        logger.debug(testrun)
+        logger.debug(testrun.metadata)
+        logger.debug(testrun.metadata.build_name)
+        build_name = testrun.metadata.build_name
+        build = testrun.build
+    return build, build_name, testrun
+
+
+def find_reproducer(group, project, device_name, debug, build_names):
+    """
+    Given a group, project, device and accepted build names, return a
+    reproducer for a test run that meets these conditions.
+    """
+    suite_name = "ltp-syscalls"
+    if debug:
+        logger.setLevel(level=logging.DEBUG)
+
+    base_group = Squad().group(group)
+    if base_group is None:
+        logger.error(f"Get group failed. Group not found: '{group}'.")
+        raise ReproducerNotFound
+
+    base_project = base_group.project(project)
+    if base_project is None:
+        logger.error(f"Get project failed. project not found: '{project}'.")
+        raise ReproducerNotFound
+
+    logger.debug(f"build name {build_names}")
+
+    metadata = first(Squad().suitemetadata(suite=suite_name, kind="test"))
+
+    if metadata is None:
+        logger.error(f'There is no suite named "{suite_name}"')
+        raise ReproducerNotFound
+
+    # == get a build that contains a run of ltp-syscalls ==
+
+    # Get the latest 10 builds in the project so we don't pick something old
+    build_options = base_project.builds(count=10, ordering="-id")
```

Comment:
Happy to change this naming - picked `build_options` to try and differentiate naming a bit since we might not find a `build` (that meets our criteria) in the X latest `builds` (which may or may not meet our search criteria).
Also happy to add a parameter for how many builds to look through (but it might be the case that this search can be simplified with some minor SQUAD changes).

Reply:
yup please do that.

Reply:
Done.